artiq/master/worker_db.py

```diff
@@ -180,7 +180,11 @@ def _write(group, k, v):
     # Add context to exception message when the user writes a dataset that is
     # not representable in HDF5.
     try:
-        group[k] = v
+        group[k] = v["value"]
+        group[k].attrs["unit"] = v["unit"]
+        group[k].attrs["scale"] = v["scale"]
+        if v["ndecimals"]:
```

Comment:
`None` values cannot be added as attributes to the HDF5 file. Unit should never be `None`, it will be an empty string instead `""`. The motivation behind allowing `ndecimals` to take a `None`  value is that it can mean not formatting the number of decimal places in the GUI. However it causes an issue when writing to the HDF5 file. It should however be possible to write to file if `ndecimals` is 0, so I will refactor to account for that case.

Reply:
Replacing the comparisons with `is None` syntax :+1: 

Leaving the value of `ndecimals` as `None` means that the value of the dataset is passed straight into `str()` (assuming it is not a numpy array). Obviously this may be ugly for some values such as `0.33333...` but I left it as an option in case the user is not concerned about precision for the dataset. Default value for `ndecimals` is still `2` in order to discourage this behavior. 

Examples:

| dataset value | formatted output |
|--------|--------|
| 1/3 | 0.3333333333333333 |
| 1.23 | 1.23 |
| 1 | 1 |

Short answer: setting `ndecimals=None` leaves floating point formatting to the `str()` function.

Can remove if this is not desired, which would enforce using a fixed number of decimal places for datasets.